### PR TITLE
a2b_ord4 OOP to AGrid2BGridFourthOrder

### PR DESCRIPTION
**Pull request by rheacangeo**
## Purpose

a2b_ord4 is done in a class AGrid2BGridFourthOrder, as everything going this way. Some attempt was made to reduce the stencil count, but the single stencil version with regions does not validate. We could do even more here, but many many many attempts to join more into the big stencil ran into validation issues with the gt backends. I suggest that be done in another PR, or wait for regions to work for the performance backends and just go to the single stencil version from here, after seeing if this helps performance at all. 
UPDATE: this slowed performance 2-3 seconds until I split up some of the stencils I had merged, may be the regions issue. No speedup gained, but no major slowdown either in testing, 

## Code changes:

- a2b_ord4 is an object -- since it gets called from divergence_damping which has not had its k_subset loop lowered, it is not 3d here, and there is a different object for each k_range
- a2b_ord4 methods have been merged until stubborn validation issues creeped back in. 
- nh_p_grad has a few different a2b objects that vary by slight krange differences and whether or not replace is True. 

## Checklist
Before submitting this PR, please make sure:

- [x] You have followed the coding standards guidelines established at [Code Review Checklist](https://paper.dropbox.com/doc/Code-Review-Checklist--BD7zigBMAhMZAPkeNENeuU2UAg-IlsYffZgTwyKEylty7NhY).
- [x] Docstrings and type hints are added to new and updated routines, as appropriate

### CODE REVIEW

**Comment by rheacangeo** [comment]
BLARG, this slows performance by 2-3 seconds per model timestep

File: fv3core/stencils/nh_p_grad.py [hunk 1]
```diff
@@ -130,6 +130,24 @@ def __init__(self):
             origin=self.orig,
             domain=self.v_domain,
         )
+        self.a2b_k1 = AGrid2BGridFourthOrder(
```
Comment (eddie-c-davis) [inline fv3core/stencils/nh_p_grad.py:133]: Super nice touch, different objects for different behavior.

**Review by eddie-c-davis — APPROVED** [review]
Epic refactor, well done! Couple of small comments...

File: fv3core/stencils/a2b_ord4.py [hunk 2]
```diff
@@ -323,337 +312,283 @@ def qout_y_edge(
...
+        3.0 * (qin[0, -1, 0] + g_in * qin) - (g_in * qy[0, 1, 0] + qy[0, -1, 0])
+    ) / (2.0 + 2.0 * g_in)
+
+
+class AGrid2BGridFourthOrder:
+    """
+    Fortran name is a2b_ord4, test module is A2B_Ord4
+    """
+
+    def __init__(self, namelist, kstart=0, nk=None, replace=False):
```
[the remark this answers is not in the document]
Comment (rheacangeo) [reply fv3core/stencils/a2b_ord4.py:396]: I didn't think we needed them on all functions?

File: fv3core/stencils/a2b_ord4.py [hunk 3]
```diff
@@ -323,337 +312,283 @@ def qout_y_edge(
...
-        qin,
-        qout,
-        grid().agrid1,
-        grid().agrid2,
-        grid().bgrid1,
-        grid().bgrid2,
-        origin=(grid().is_, grid().js, kstart),
-        domain=corner_domain,
-    )
+    def compute_qout_edges(self, qin, qout):
```
[the remark this answers is not in the document]
Comment (rheacangeo) [reply fv3core/stencils/a2b_ord4.py:565]: I thought we made the stencil wrapper attribute private to that class, but not the definition of the function. It has probably varied throughout the repo

File: fv3core/stencils/a2b_ord4.py [hunk 4]
```diff
@@ -323,337 +312,283 @@ def qout_y_edge(
...
+        self._a2b_interpolation_stencil = StencilWrapper(
+            a2b_interpolation, externals=ax_offsets, origin=origin, domain=domain
+        )
+        self._copy_stencil = StencilWrapper(
+            copy_stencil_method,
+            origin=(self.grid.is_, self.grid.js, kstart),
+            domain=(self.grid.nic + 1, self.grid.njc + 1, nk),
+        )
+
+    def __call__(self, qin, qout):
```
[the remark this answers is not in the document]
Comment (rheacangeo) [reply fv3core/stencils/a2b_ord4.py:499]: I can add them, didn't really think we needed them everywhere, but is easy to do

File: fv3core/stencils/a2b_ord4.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rheacangeo) [reply fv3core/stencils/a2b_ord4.py:565]: oh this one, you are right, nevermind

File: fv3core/stencils/a2b_ord4.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rheacangeo) [reply fv3core/stencils/a2b_ord4.py]: added!

**Pull request merged**